scripts/linkchecker.py

```diff
@@ -0,0 +1,395 @@
+    AND we check if 'foo/bar/_index.md' exists.
+
+    :param base: The base directory to begin with
+    :param path: The link target which is a relative path string
+    :returns: A boolean indicating whether the target file exists.
+    """
+    # NOTE: anchor is ignored, can be a todo item
+    parts = path.split("#")
+
+    fn = normalize_filename(parts[0], ftype=ftype)
+    target = base + fn
+
+    if os.path.isfile(target):
+        return True
+
+    dir_name = os.path.join(base, parts[0])
+    if os.path.isdir(dir_name):
+        target = os.path.join(dir_name, "_index.md")
+        if os.path.isfile(target):
+            return True
+    return False
+
+
+def is_redirect(path):
+    """Check if the path exists in the redirect database.
+
+    NOTE: We do NOT check if the redirect target is there or not. We do an
+    **exact** matching for redirection entries.
+    """
+    global REDIRECTS
+
+    # NOTE: anchor is ignored, can be a future todo
+    parts = path.split("#")
+    target = parts[0]
+    if not target.endswith("/"):
+        target += "/"
+
+    for item in REDIRECTS:
+        if item == target:  # EXACT MATCH
+            return True
+    return False
+
+
+def check_target(page, anchor, target):
+    """Check a link from anchor to target on provided page.
+
+    :param page: Currently not used. Passed here in case we want to check the
+                 in-page links in the future.
+    :param anchor: Anchor string from the content page. This is provided to
+                help handle cases where target is empty.
+    :param target: The link target string to check
+    :returns: A checking record (string) if errors found, or None if we can
+              find the target link.
+    """
+    # B01: bad protocol
+    if target.startswith("http://"):
+        return new_record("ERROR", "Bad protocol", target)
+
+    # full link
+    if target.startswith("https://"):
+        # B03: self link, should revise to relative path
+        if (target.startswith("https://k8s.io/website/docs") or
+                target.startswith("https://kubernetes.io/website/docs")):
+            return new_record("ERROR", "Should use relative paths", target)
+        # external link, skip
+        return new_record("INFO", "External link", target)
+
+    # in-page link
+    # TODO: check if the target anchor does exists
+    if target.startswith("#"):
+        return new_record("INFO", "In-page link, skipped", target)
+
+    # Link has shortcode
+    if target.find("{{") > 0:
+        return new_record("WARNING", "Link has shortcode, skipped", target)
+
+    # link to English or localized page
+    if (target.startswith("/docs/") or
+            target.startswith("/" + ARGS.lang + "/docs/")):
+        # it is an embedded image
+        # TODO: an image might get translated as well
+        if target.endswith(".png"):
+            return new_record("INFO", "Link to image, skipped", target)
```

Comment:
Hi @tengqm .
I am reading through the script. Quick work. Some feedback:
Is the idea to replace or use along with the `wjdp/htmltest` link checker?
The goal of the script is to report/identify incorrect links throughout the docs portion of the site (no blog) compared
against the actual content directory structure and the _redirects file?
Will the script provide any insight into as to how to correct the link or what the link should be?
Do you want to check images with a suffix of `.svg`?
I have not checked, does the regex return _unique_ paths on the page or all paths?
I will test the script.

Reply:
yes. the checker checks all langs, including redirects.
it doesn't provide suggestions yet, though can be added.
right, svg should be checked as well, will add.
I think it returns all paths.